AIPS/aip-18.md

```diff
@@ -0,0 +1,205 @@
+      "nonce": "String",
+      "resolution": [Float, Float]
+    }
+  ],
+  "storage": { "id": "String" },
+  "comment": "String"
+}
+````
+
+Object fields:
+
+- `files`: An array containing information about the attached encrypted files
+- `id`: Represents the unique file identifier in the `storage`
+- `type`: Denotes the type of the file after decryption. Optional.
+- `size`: Indicates the size of the encrypted file in bytes
+- `preview`: Image preview. Optional.
+- `name`: Specifies the name of a file. Optional.
+- `nonce`: Nonce used for encryption
+- `storage`: Specifies a decentralized storage information where a file is stored
+- `comment`: A comment/caption associated with a file. Optional.
+- `resolution`: File resolution as an array of float values, where the first value denotes the width and the second value denotes the height of a file. Useful for images and videos. Optional.
+
+### Examples
+
+Object `transaction.asset.chat.message` *before encryption*, which includes a single file:
+
+```` json
+{
+  "storage": {
+    "id": "ipfs"
+  },
+  "files": [
+    {
+      "id": "bafkreifxs2buan4anxwhfnt...v25rewteppsh6u4s4x5hi4o4du",
+      "name": "IMG_308",
+      "resolution": [
+        1179,
+        2556
+      ],
+      "preview": {
+        "id": "bafkreidqpnqmqocqp...if4xb3bi76gl6g4kmtg62cmpw5ufu",
+        "nonce": "297d443aa29aa...485309f9f5ae159df9784d"
+      },
+      "nonce": "030aa714943dc1dd...1c7091920d9d",
+      "type": "png",
+      "size": 150998
+    }
+  ],
+  "comment": "Hello!"
+}
+````
+
+The entire transaction after encryption from U15677078342684640219 to U7972131227889954319 looks like this:
+
+```` json
+{
+  "transaction": {
+    "type": 8,
+    "amount": 0, // Amount should be 0
+    "senderId": "U15677078342684640219",
+    "senderPublicKey": "e16e624fd0a5123294b448c21f30a07a0435533c693b146b14e66830e4e20404",
+    "asset": {
+      "chat": {
+        "message": "70cbd07ff2ceaf0fc38a01ef9...", // files object encrypted
+        "own_message": "e98794eaedf47e...",
+        "type": 2 // Type should be 2 (Rich Content Message)
+      }
+    },
+    "recipientId": "U7972131227889954319",
+    "timestamp": 46116887,
+    "signature": "8fc2a54604109a6fcdccec2..."
+  }
+}
+````
+
+You can combine sending files with [AIP-16](https://aips.adamant.im/AIPS/aip-16)) to reply to a message with a file attached. Send the following structure in the `reply_message` field:
+
+```` json
+{
+  "replyto_id": "7452709338464950789",
+  "reply_message": {
+    "storage": {
+      "id": "ipfs"
+    },
+    "files": [
+      {
+        "id": "bafkreifxs2buan4anxwhfnt...v25rewteppsh6u4s4x5hi4o4du",
+        "name": "IMG_308",
+        "resolution": [
+          1179,
+          2556
+        ],
+        "preview": {
+          "id": "bafkreidqpnqmqocqp...if4xb3bi76gl6g4kmtg62cmpw5ufu",
+          "nonce": "297d443aa29aa...485309f9f5ae159df9784d"
+        },
+        "nonce": "030aa714943dc1dd...1c7091920d9d",
+        "type": "png",
```

Comment:
Instead of "type" it would be better to store "mimeType" and "extension". This way there will be no need to extract the MIME from the extension. It is not always possible to get the correct MIME-type from an extension. In addition, a lot will depend on how the extractor is implemented on different clients. We will encounter the fact that the same file can have different types on different clients. To avoid this, I suggest adding this information to Adamant’s message, because it will be much more convenient.